### PR TITLE
HACKING.md,CODING.md: add a coding/review checklist for the project 

### PR DESCRIPTION
**Pull request by pedronis**
This captures considerations and practices that we have evolved in the project and should be taken into account when code is reviewed for it.

### CODE REVIEW

File: CODING.md [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+  Ideally all exported names should have doc comments for them following [golang conventions](https://go.dev/doc/comment).
+
+We sometimes also use long code comments or separate markdown README files for higher-level descriptions of mechanisms or concepts.
+
+Inline code comments should usually address non-obvious, unexpected parts of the code. Repeating what the code does is not usually very informative:
+
+* Code comments should either address the why something is done
+* Or clarify the more abstract impact of the low-level manipulation in the code
+
+ It might be appropriate and useful also to give proper doc comments even to complex unexported helpers.
```
Comment (degville) [inline CODING.md:34]:
```suggestion
It might be appropriate and useful also to give proper doc comments even to complex unexported helpers.
```

Comment (degville) [reply CODING.md:34]: Just to avoid this likely being indented alongside the above list items.

File: CODING.md [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+* OTOH as error messages often end up being embedded in one another/chained. It is also important to pay attention so that the final messages do not have too much repetition, when possible, to avoid things like *"cannot …: cannot …: cannot …"* for example. Tests for the error paths can help find those repetitions.
+
+* Error messages should be clear, and when possible, actionable. They should also use concepts and terminology familiar to the user instead of internal-only concepts unless they are really unexpected internal errors.
+
+* Prefixing errors with *"internal error: …"* should be used for programming errors or other unexpected internal inconsistencies but preferably not for situations where external state that is not completely under snapd control is involved.
+
+## Other style points
+
+* We rely on and apply `go fmt` consistently.
+* Our PR CI static checks run [`golangci-lint`](https://github.com/golangci/golangci-lint), for details see our .golangci-lint.yml config
```
Comment (degville) [inline CODING.md:72]:
```suggestion
* Our PR CI static checks run [`golangci-lint`](https://github.com/golangci/golangci-lint), for details see our `.golangci-lint.yml config`.
```

**Review by degville — APPROVED** [review]
Looks good. Thank you for writing this.

File: CODING.md [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+     timeNow = f
+     return restore
+}
+```
+
+If something cannot avoid being mocked across package boundaries, we sometimes have `Mock*` functions or constructors exported in the API of packages.
+
+Because of this complexity with mocking, and because mock-heavy tests might risk needing large rewrites when refactoring (which goes against their confidence enhancement use) we are not very strict about unit tests testing exactly single functions and structs. We should do that whenever possible without mocking, but otherwise it is not atypical for snapd tests to concentrate on mocking points of interaction with the actual external system and state, as it might require less overall mocking support and it might be easier to reason on expectations for effects. For example, we have support to mock our systemd interactions and observe the involved `systemctl` invocations (`systemd.MockSystemctl`).
+
+So, many of our unit tests might end up testing more than one package, and test instead across two levels (rarely more) of packages in our architecture; a lower-level primitives' package, for example, and a more high-level behaviour one using the former.
```
Comment (MiguelPires) [inline CODING.md:160]:
```suggestion
So, many of our unit tests might end up testing more than one package, and test instead across two levels (rarely more) of packages in our architecture; a lower-level primitives package, for example, and a more high-level behaviour one using the former.
```

File: CODING.md [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+
+Full direct mocking might still make perfect sense when the API of the consumed packages is very complex but its details should be fully or largely transparent to the consumer. This is mostly the case, for example, when testing API functionality in the `daemon` package vs the API offered by the [overlord state managers](https://github.com/snapcore/snapd/blob/master/overlord/README.md).
+
+The cost of our approach is sometimes a complex fixture setup. To help mitigate this, and in other cases when it makes sense for a package to offer test-dedicated helpers related to it, we can introduce matching `<main-package>test` packages one level deeper than `<main-package>` (e.g. `asserts/assertstest` or `overlord/devicestate/devicestatetest`).
+
+Related to tests in and for overlord state manager packages `overlord/<concern>state`, we have a few rules:
+
+* Ideally they should limit themselves to test the manager defined by the package
+* If that's not possible they should limit themselves to as few managers as possibles
```
Comment (MiguelPires) [inline CODING.md:169]:
```suggestion
* If that's not possible they should limit themselves to as few managers as possible
```

File: CODING.md [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+* If what needs to be tested is the full interaction across many or all managers then we have or can write tests for this in overlord/managers_test.go. Fixture setup in these cases is very costly but they are still easier to iterate on and can be useful to probe behaviour in more internal details than functional/integration tests.
```
Comment (MiguelPires) [inline CODING.md:170]:
```suggestion
* If what needs to be tested is the full interaction across many or all managers then we have or can write tests for this in `overlord/managers_test.go`. Fixture setup in these cases is very costly but they are still easier to iterate on and can be useful to probe behaviour in more internal details than functional/integration tests.
```

File: CODING.md [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+
+We do not have strong policies against repetition in test code, as usual the important consideration is readability. This area is mostly left to the personal judgement of developers. If any general advice can be given is that:
+
+* investing in clear helpers to setup complex fixtures is often valuable, while compressing actual ad hoc testing and checking code less so, as it might result in if-trees that might be hard to follow
+* wherever applicable, tabular tests (where cases are expressed as a slice of anonymous structs) should be used. For example, they are often appropriate when testing error cases of functions
```
Comment (MiguelPires) [inline CODING.md:175]: The rest of the document capitalises first words in lists
```suggestion
* Investing in clear helpers to setup complex fixtures is often valuable, while compressing actual ad hoc testing and checking code less so, as it might result in if-trees that might be hard to follow
* Wherever applicable, tabular tests (where cases are expressed as a slice of anonymous structs) should be used. For example, they are often appropriate when testing error cases of functions
```

File: CODING.md [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+
+* PR summaries are expected to be of this form:
+  * *`affected full packages:  short summary in lowercase`*
+    * When too many packages are involved, many can be used instead, or sometimes package names can be abbreviated by using single letters for the top-level package, when non ambiguous combined with the subpackage.
+    * Examples:
+      * `overlord/devicestate: add test to check connect hooks don't break anything`
+      * `gadget,image: remove LayoutConstraints struct`
+      * `o/snapstate: add helpers to get user and gating holds`
+      * `many: correct struct fields and output key`
+  * When no go code is involved, the context prefix before the colon can refer to directories or top-level files instead.
```
Comment (MiguelPires) [inline CODING.md:201]:
```suggestion
  * When no Go code is involved, the context prefix before the colon can refer to directories or top-level files instead.
```

**Review by MiguelPires — COMMENTED** [review]
Looks good, thanks for this. Most of my comments are about minor typos or suggestions. One thing that I didn't point out in the diff is that list item sentences terminate with periods in a seemingly inconsistent way (most do, some don't). I'm not sure if you care, I'm fine with it since it's not a formal paper or something, but just thought I'd mention it in case you care

File: CODING.md [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+
+* Code can have a long lifetime; the effort to maintain and adapt it in the future can be much larger than the original effort to produce the first version of it. Reviews from other team member should therefore focus on:
+  * Is the new code readable and understandable, alongside other attributes that can help future maintainability?
+  * Could the code be simplified?
+
+## Naming conventions
+
+To a large extent we follow [golang naming conventions](https://go.dev/doc/effective_go#names):
+
+* Names should strike a balance between concision and clarity, where for a local variable more weight might be put on concision while for an exported name clarity might have a larger weight.
```
[the remark this answers is not in the document]
Comment (pedronis) [reply CODING.md:15]: I would prefer the keep the spelling, it's also the main spelling chosen by wikipedia for this (conciseness redirects to concision).

File: CODING.md [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+
+* When trying to keep functions and methods readable by introducing helpers, trying to aim for a mostly consistent level of abstraction inside each function could be useful.
+
+* *Do not repeat yourself* is a balancing act. Complex behaviour should ideally be encoded in code only once in the code base when possible. When extracting and deciding how to extract some behaviour it is important to consider the readability of both the now encapsulated code and its consumers. For example, if it's hard to give the extracted code a good name and signature it might show that a different approach should be looked at. For simpler helpers, it might be worth seeing a couple of usages before creating them as local helpers, and a bit more before creating an exported helper that can be imported from all the used places. When creating helpers and avoiding repetition the aim should also be first to improve maintainability and readability. If the consumer code is less readable then maybe the extraction in this case might not be a good idea in the end.
```
[the remark this answers is not in the document]
Comment (pedronis) [reply CODING.md:88]: I removed the "in code" bit instead

File: CODING.md [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+
+## Tests
+
+Tests should help with these aspects:
+
+* Illustrate (and help clarify while writing the code) the intent and behaviour of code.
+* Anchor the correctness of the code, to the extent possible, from this POV while we try to keep a high coverage without overdoing. One should keep in mind that line coverage might itself not be enough. Lines of code can be covered without being tested as only part of a predicate might be triggered, or some code could be removed and no test breaks unless tests were targeting it. So when writing tests it is always important to take on an expectations and behaviours mindset.
+* Help later to have some confidence when refactoring.
+
+We do not mandate TDD, but it's always a good idea when possible to start at least from happy path tests for new code/behaviour.
```
[the remark this answers is not in the document]
Comment (pedronis) [reply CODING.md:122]: I add this sentence

> Bug fixes should be covered by new tests, for which is important to verify that they do not pass and fail as expected prior to the fix.

File: CODING.md (hunk 7 above)
Comment (pedronis) [reply CODING.md:201]: I used golang as the rest of the document

File: CODING.md [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+
+## Comments
+
+  Ideally all exported names should have doc comments for them following [golang conventions](https://go.dev/doc/comment).
+
+We sometimes also use long code comments or separate markdown README files for higher-level descriptions of mechanisms or concepts.
+
+Inline code comments should usually address non-obvious, unexpected parts of the code. Repeating what the code does is not usually very informative:
+
+* Code comments should either address the why something is done
```
[the remark this answers is not in the document]
Comment (pedronis) [reply CODING.md:31]: kept as is

File: CODING.md (hunk 3 above)
Comment (pedronis) [reply CODING.md:160]: I switched to a "package of" construction

**Review by MiguelPires — APPROVED** [review]
LGTM, thanks

**Comment by pedronis** [comment]
force pushed to have a slightly cleaner history
